heartbeat/eventext/eventext.go

```diff
@@ -0,0 +1,32 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package eventext
+
+import (
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// MergeEventFields merges the given common.MapStr into the given Event's Fields.
+func MergeEventFields(e *beat.Event, merge common.MapStr) {
+	if e.Fields != nil {
+		e.Fields.DeepUpdate(merge)
+	} else {
+		e.Fields = merge
+	}
+}
```

Comment:
I didn't check if that's the case, but in general be always cautious when sharing `common.MapStr` via DeepUpdate or assigning to e.Fields. They might be modified by processors, potentially leading to panics and crashes due to races. If processors only apply to a subset of events you might also end up with invalid event fields. When in doubt call `MergeEventFields(e, fields.Clone())`.

Reply:
Btw. I'm aware you moved it :) Just wanted to point a common programming error we've seen in the past in heartbeat and a few other places. Maybe we should add a code comment telling users of this function to use this only with freshly constructed MapStrs or a Clone of an existing MapStr.

Reply:
Yes, I've checked this in the past. Ideally we'd enforce this via code instead of convention, but I'm not worried about clones at the moment.